### PR TITLE
[catalog] Make unprocessed entities visible

### PR DESCRIPTION
**Pull request by alde**
Create an API route to fetch failed and pending entities, as well as a frontend plugin to view them.

To clarify, an Unprocessed entity has either failed processing or is pending to be processed. It's been registered in Catalog but "can't be found". Currently the only way to find why it hasn't processed is to check the logs.

Added a broken entity to the sample data.

Opening this up for discussion.

Screenshot 2023-05-16 at 3 38 35 PM
Screenshot 2023-05-16 at 3 56 48 PM

### CODE REVIEW

File: plugins/catalog-unprocessed-entities/src/plugin.ts
Comment (awanlin) [inline plugins/catalog-unprocessed-entities/src/plugin.ts:54]: Could we also maybe export a hypothetical `CatalogUnprocessedEntitiesContent` that would be just the content of this page so that is could be added as a tab in the new DevTools plugin? Or if you are open to it maybe all the frontend would just live there as well? Your call but the first one would be super nice!

File: plugins/catalog-unprocessed-entities/src/components/EntityDialog.tsx
Comment (awanlin) [inline plugins/catalog-unprocessed-entities/src/components/EntityDialog.tsx:63]: Would it make more sense to use the `CodeSnippet` component here from `@backstage/core-components` instead? That's what the Entity Inspector does to show the raw JSON or YAML: https://github.com/backstage/backstage/blob/2d98053d5264f7b86c05b4a80024afc41fe33012/plugins/catalog-react/src/components/InspectEntityDialog/components/JsonPage.tsx#L32

**Review by awanlin — COMMENTED** [review]
Thanks for this @alde, feel like many in the Community, as well as Adopters, will be very happy to see this get merged in eventually 🚀

File: plugins/catalog-unprocessed-entities/src/components/EntityDialog.tsx
Comment (alde) [reply plugins/catalog-unprocessed-entities/src/components/EntityDialog.tsx:63]: I didn't know about that, that would make more sense. WIll look into it.

File: plugins/catalog-unprocessed-entities/src/plugin.ts
Comment (alde) [reply plugins/catalog-unprocessed-entities/src/plugin.ts:54]: I'm open to exporting more things, not sure exactly what you mean though

Comment (awanlin) [reply plugins/catalog-unprocessed-entities/src/plugin.ts:54]: Ok, so with version `1.14.0` we introduced the DevTools plugin which acts as a place to collect tooling that helps with the Ops side of Backstage - shows Backstage version, package versions, current config, etc. DevTools can be customized to add more tabs and being able to add this there would be handy, here's the docs: https://github.com/backstage/backstage/tree/master/plugins/devtools#customizing

Now, if we could break out the `<Content>` section of `UnprocessedEntities` into it's own component and then export that in the `plugin.ts` it could be used as a tab in DevTools. I think you may not even need to export it in the `plugin.ts` but just export it like any other component

Hopefully that better explains what I'm asking for?

File: packages/catalog-model/examples/components/invalid-component.yaml
Comment (nuritiz) [inline packages/catalog-model/examples/components/invalid-component.yaml:4]:
```suggestion
  name: invalid component  # wrong name convention 
```

Comment (alde) [reply packages/catalog-model/examples/components/invalid-component.yaml:4]: I did this with the description of the entity rather than as a commen.

File: plugins/catalog-backend/src/catalog/types.ts
Comment (benjdlambert) [inline plugins/catalog-backend/src/catalog/types.ts:228]: We should probably wrap this up into an object instead of returning just an array, much better for evolving the API over time without breaking the response type if we want to add another field or something.

File: plugins/catalog-backend/src/service/AuthorizedEntitiesCatalog.ts
Comment (benjdlambert) [inline plugins/catalog-backend/src/service/AuthorizedEntitiesCatalog.ts:269]: I'm wondering if we should make this whole extension a module that you can install into the catalog instead of building it directly into the catalog backend plugin. It would make things simpler in the catalog-backend and encourage people to start breaking out these extensions into modules instead.

**Review by benjdlambert — CHANGES_REQUESTED** [review]
Some quick comments about some higher level things - cc @backstage/maintainers as to what we think would be a good path forward here for adding more functionality to the catalog like this?

File: plugins/catalog-backend/src/service/AuthorizedEntitiesCatalog.ts
Comment (Rugvip) [reply plugins/catalog-backend/src/service/AuthorizedEntitiesCatalog.ts:269]: Yep I think module is the best approach here idd. A plain `createBackendModule` for the new system, and for the old one a separate `createUnprocessedEntitiesRouter` that you install alongside the main router in the catalog backend.

Comment (alde) [reply plugins/catalog-backend/src/service/AuthorizedEntitiesCatalog.ts:269]: My first idea was to add an `addAdditionalRoutes` function to the CatalogBuilder which I think would be cleaner, but the problem with that was that the EntitiesCatalog gets built in the Builder. Since it's needed in the route handlers I'm not sure how to solve the dependency injection.

File: plugins/catalog-backend-module-unprocessed/src/UnprocessedEntitiesModule.ts
Comment (benjdlambert) [inline plugins/catalog-backend-module-unprocessed/src/UnprocessedEntitiesModule.ts:107]: One refactor I would probably suggest here, is to make this function called `getRouter` and then you can return a `Router()` instance in here and then use it in both the module by using `router.use(module.getRouter())` and the same for the old system too. Makes it a little cleaner, and it means less dependencies injected into the contructor.

File: packages/backend-next/src/index.ts
Comment (alde) [inline packages/backend-next/src/index.ts:64]: broken at the moment, should be fixed by https://github.com/backstage/backstage/pull/17953

Comment (benjdlambert) [reply packages/backend-next/src/index.ts:64]: Rebased and verified as working! :tada:

**Comment by leon-vg** [comment]
I would love to have an added feature of detecting duplicate entries. We see some copy pasting going on of `catalog-info.yaml` files (starting a repo from the basis of another), and the `catalog-info.yaml` is not updated consequently.

These files don't seem to have processing errors, but they fail when the entries are saved to the database, due to unique constraints.

The query we use to get these duplicates is this one, but I'm sure there must be a better way somehow.

```
await dbClient<Columns>('refresh_state')
    .select('*')
    .leftJoin(
      'refresh_state_references',
      'refresh_state_references.source_entity_ref',
      'refresh_state.entity_ref',
    )
    .where('errors', '[]')
    .whereLike('refresh_state.entity_ref', 'location:%')
    .whereNot('refresh_state.cache', '{"ttl":5,"cache":{}}')
    .whereNull('refresh_state_references.source_entity_ref');
```

**Comment by freben** [comment]
Can't really detect dupes/conflict with a db query since at the point in time that the conflict happens, the conflicting information is thrown away.

I've been experimenting with an iteration on the db schema where every ingestion is retained for posterity, but nothing that'll happen in the near term.

**Comment by leon-vg** [comment]
In the above query I mentioned the trick is to check for a cache record that the urlReader leaves behind in the `refresh_state` table. It seems to work in our context, but I guess it's not a generic approach then.

**Pull request merged**
